### PR TITLE
fix ascii aliases and url skipping

### LINKED ISSUE #54: emoji detection too eager?
`:heart:oops` becomes `❤️open_mouth:ops`
`https://foo:oops@example.com` becomes `https://foo😮ops@example.com`

See also jitsi/jitsi-meet#6501.

### PR DESCRIPTION
**Pull request by obfusk**
* replace full aliases before (and after) ascii aliases
  (fixes e.g. ":heart:oops")
* skip urls entirely (fixes e.g. "https://foo:oops@example.com")

fixes #54

### CODE REVIEW

**Comment by enzoferey** [comment]
Thanks for opening @obfusk. Could you please add these cases to the test suite and check all tests are passing ? Right now if I run your changes locally I get 5 non passing test. We don't have any CI in this project at the moment unfortunately.

File: src/asciiRegex.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ const names = flatten(
   })
 ).join("|");
 
-const edgeCases = ["http", "https"].join("|");
+const edgeCases = "https?:\\S*";
```
Comment (enzoferey) [inline src/asciiRegex.js:13]: We want to keep this an array and support both http and https URLs.

Comment (obfusk) [reply src/asciiRegex.js:13]: Since the `?` in `https?` makes the `s` optional, it does support both http and https URLs.

**Comment by obfusk** [comment]
> Could you please add these cases to the test suite and check all tests are passing ? Right now if I run your changes locally I get 5 non passing test. We don't have any CI in this project at the moment unfortunately.

I was hoping CI would catch any mistakes I made :sweat_smile:
I fixed the code for the tests that should not have broken.
I changed one test (`https://google.com:)`) for which the emoji is now considered part of the url. I personally think this is OK since there is no completely reliable -- and certainly no simple -- regex for matching only valid urls before an emoji.
And I added some CI on another branch: https://github.com/obfusk/react-emoji-render/tree/fix-regexes-and-ci

File: src/asciiRegex.js (hunk 1 above)
Comment (enzoferey) [reply src/asciiRegex.js]: Oh, definitely 🤦

File: src/asciiRegex.js [hunk 2]
```diff
@@ -10,7 +10,7 @@ const names = flatten(
   })
 ).join("|");
 
-const edgeCases = ["http", "https"].join("|");
+const edgeCases = "https?\\S*";
```
Comment (enzoferey) [inline src/asciiRegex.js:13]: Could you explain me why matching against `\` and `S` literally ?

File: src/__tests__/index-test.js [hunk 3]
```diff
@@ -175,10 +179,14 @@ describe("toArray", () => {
...
   });
-  test("url and emoji (no space)", () => {
+  test("url and ascii emoji alias (no space)", () => {
     const content = toArray("https://google.com:)");
     expect(content).toMatchSnapshot();
   });
+  test("url and emoji (no space)", () => {
+    const content = toArray("https://google.com:heart:");
+    expect(content).toMatchSnapshot();
+  });
```
Comment (enzoferey) [inline src/__tests__/index-test.js:189]: These 2 tests should be consistent. Meaning that is think `:)` is part of the URL and thus not matching, `:heart:` should not be matching neither (which is not the case). I think indeed we should see this case as part of the URL as create no match.

File: src/asciiRegex.js (hunk 2 above)
Comment (obfusk) [reply src/asciiRegex.js:13]: It's not matching literally. There are two backslashes in the string literal because literal backslashes must be escaped there. The string is then converted to a RegExp, where `\S` matches any non-whitespace character.

```javascript
> RegExp("https?\\S*")
/https?\S*/
> RegExp("https?\\S*").exec("https://example.com/foo/bar/baz :)")[0]
'https://example.com/foo/bar/baz'
```

File: src/__tests__/index-test.js (hunk 3 above)
Comment (obfusk) [reply src/__tests__/index-test.js:189]: Let's see what github does:
* `https://example.com:)` https://example.com:)
* `https://example.com/foo/:heart:` https://example.com/foo/:heart:
* `https://example.com/foo/:heart:/bar` https://example.com/foo/:heart:/bar
* `https://example.com/foo:heart:bar` https://example.com/foo:heart:bar

Not exactly ideal either :(

I think the most important situation, which this PR fixes, is the case of short ascii aliases like `:p`, which you really don't want to be replaced in random places in URLs. I think the "full names" like`:heart:` are less likely to cause problems, so I'd be OK with either approach for those.

That would of course need to be fixed in another part of the code:

```javascript
const urls = "https?\\S*";
function getAliasesRegex() {
  return new RegExp(`(?<!${urls}):([${allowedAliasCharacters}]+):`, "g");
}
```

I can add that change to the PR if you want.

File: src/asciiRegex.js (hunk 2 above)
Comment (enzoferey) [reply src/asciiRegex.js:13]: Definitely dumb ! 🤣 Forgot about string context, sorry. You are totally right, this is the right approach 👍

File: src/__tests__/index-test.js (hunk 3 above)
Comment (enzoferey) [reply src/__tests__/index-test.js:189]: Right, I think there will always be people on one side of the other. I think for the sake of simplicity in our case and be consistent across all possible edge cases, we should not replace any ascii or full name alias that is part of a URL. Meaning, all the cases above should not convert. If you want to place an emoji next to a URL, you can do it before (with or without space), or after (with space).

How does that sound to you ? cc @tommoor as well

Comment (obfusk) [reply src/__tests__/index-test.js:189]: Fine with me. I've added a commit for that.

File: src/__tests__/index-test.js [hunk 4]
```diff
@@ -175,14 +179,22 @@ describe("toArray", () => {
...
   test("url and emoji (with space)", () => {
```
Comment (enzoferey) [inline src/__tests__/index-test.js:190]: Let's rename this one so it has a different name than the one you added just below.

```suggestion
  test("url and ascii emoji (with space)", () => {
```

File: src/renderer.js [hunk 5]
```diff
@@ -80,7 +80,7 @@ export function toArray(text, options = {}) {
         }
 
         const isMaybePartOfBiggerAlias =
-          maybeBiggerAliasCharacters !== undefined;
+          maybeBiggerAliasCharacters !== undefined && fullMatch[0] == ":";
```
Comment (enzoferey) [inline src/renderer.js:83]: Why do you need this ? After removing it tests still pass.

File: src/renderer.js [hunk 6]
```diff
@@ -116,6 +116,9 @@ export function toArray(text, options = {}) {
     while (previousTextWithoutAsciiAliases !== textWithoutAsciiAliases) {
       previousTextWithoutAsciiAliases = textWithoutAsciiAliases;
       textWithoutAsciiAliases = textWithoutAsciiAliases.replace(
+        aliasesRegex,
+        replaceAliases
```
Comment (enzoferey) [inline src/renderer.js:120]: This replace here doesn't make a lot of sense due to the context on the function. It should be executed on `text` before the call to `replaceAllAsciiAliases` (see next comment).

File: src/renderer.js [hunk 7]
```diff
@@ -124,10 +127,10 @@ export function toArray(text, options = {}) {
     return textWithoutAsciiAliases;
   }
 
-  const textWithouAsciiAliases = replaceAllAsciiAliases(text);
+  const textWithoutAsciiAliases = replaceAllAsciiAliases(text);
 
   return replace(
-    textWithouAsciiAliases.replace(aliasesRegex, replaceAliases),
+    textWithoutAsciiAliases.replace(aliasesRegex, replaceAliases),
```
Comment (enzoferey) [inline src/renderer.js:133]: Here we basically want to do this:

1. Replace aliases.
2. Replace ascii aliases to aliases.
3. Replace aliases.

I would do it something like this:

```js
let replacedText = text;
replacedText = replacedText.replace(aliasesRegex, replaceAliases);
replacedText = replaceAllAsciiAliases(replacedText);
replacedText = replacedText.replace(aliasesRegex, replaceAliases);

return replace(replacedText, unicodeEmojiRegex, replaceUnicodeEmoji);
```

**Comment by enzoferey** [comment]
Applying the changes above should still make all test pass. After thinking through this, I think the algorithm to replace aliases could be simplified quite a bit. We will leave that for another day, most important is to ship a fix for this 👍 

Thank you for the effort 🙏

File: src/renderer.js (hunk 5 above)
Comment (obfusk) [reply src/renderer.js:83]: The code assumes later on that the first character of the full match is a `:` when it checks whether `.slice(1, -1)` is an alias (e.g. `:o` and `:open_mouth:`). There are however several ascii aliases that don't start with `:`, like `<3`.

It does turn out that I made a mistake in checking.
Previously, `<3:` would not be replaced at all, this version replaces it with `❤️`, but we want `❤️:`.

If you put the check later it works:

```javascript
if (fullMatch[0] == ":" && isPartOfBiggerAlias) {
```

Comment (obfusk) [reply src/renderer.js:83]: Maybe fix it that way and add a test?

File: src/renderer.js (hunk 7 above)
Comment (obfusk) [reply src/renderer.js:133]: I was trying to prevent a problem with replaced ascii aliases interfering with the next loop of processing ascii aliases, but it's not working as I intended.

I found an example of a similar problem: `":1<3" -> ":1:heart:" -> "<EMOJI>heart:"`

I suggest changing the algorithm to:

```javascript
let replacedText = text;
replacedText = replacedText.replace(aliasesRegex, replaceAliases);
replacedText = replaceAllAsciiAliases(replacedText);
return replace(replacedText, unicodeEmojiRegex, replaceUnicodeEmoji);
```

and also change `replaceAsciiAliases`:

```javascript
if (!isMaybePartOfBiggerAlias) {
  return aliases[alias];
  // instead of `:${alias}:`
}

return `${aliases[alias]}${maybeBiggerAliasCharacters}`;
// instead of `:${alias}:${maybeBiggerAliasCharacters}`
```

Now we get `":1<3" -> ":1❤️"`.

File: src/renderer.js (hunk 6 above)
Comment (obfusk) [reply src/renderer.js:120]: (see reply to next comment)

File: src/renderer.js (hunk 7 above)
Comment (obfusk) [reply src/renderer.js:133]: If you agree, I'll make the changes and add tests for those cases.

File: src/renderer.js (hunk 5 above)
Comment (enzoferey) [reply src/renderer.js:83]: I see, yes definitely ! 🙏 Make triple equals thought 😉 (`fullMatch[0] === ":`)

(answer continues in next comment)

File: src/renderer.js (hunk 7 above)
Comment (enzoferey) [reply src/renderer.js:133]: Definitely, that one of the things I meant at https://github.com/tommoor/react-emoji-render/pull/55#issuecomment-626205705. Do whatever you think is required and please add tests for these edge cases that are really good 👍

File: src/asciiRegex.js [hunk 8]
```diff
@@ -10,8 +10,6 @@ const names = flatten(
   })
 ).join("|");
 
-const edgeCases = ["http", "https"].join("|");
```
Comment (enzoferey) [inline src/asciiRegex.js:13]: I also think I would like to keep this. Instead of just `startOrURL`, we can have `const edgeCases = ["https?\\S*"].join("|");`. That way it's ready for future other things we would like to blacklist and it matches with the business logic at https://github.com/tommoor/react-emoji-render/blob/master/src/renderer.js#L67.

**Comment by obfusk** [comment]
I think the current version of the PR resolves all issues.

**Comment by enzoferey** [comment]
@tommoor I let you publish a new version of the package 🙌

**Comment by tommoor** [comment]
`v1.2.3`

**Comment by obfusk** [comment]
Maybe update the codesandbox in the README as well?
